src/tests/simulation/tests/integration_tests/scale/auto_parallelism.rs

```diff
@@ -0,0 +1,628 @@
+    cluster
+        .simple_kill_nodes(vec!["compute-2".to_string()])
+        .await;
+
+    sleep(Duration::from_secs(
+        MAX_HEARTBEAT_INTERVAL_SECS_CONFIG_FOR_AUTO_SCALE * 2,
+    ))
+    .await;
+
+    session.run("create table t (v1 int);").await?;
+    session
+        .run("create materialized view m as select count(*) from t;")
+        .await?;
+
+    session
+        .run("insert into t select * from generate_series(1, 100)")
+        .await?;
+    session.run("flush").await?;
+
+    sleep(Duration::from_secs(5)).await;
+
+    let table_mat_fragment = cluster
+        .locate_one_fragment(vec![
+            identity_contains("materialize"),
+            no_identity_contains("simpleAgg"),
+        ])
+        .await?;
+
+    let (all_parallel_units, used_parallel_units) = table_mat_fragment.parallel_unit_usage();
+
+    assert_eq!(all_parallel_units.len(), used_parallel_units.len());
+
+    assert_eq!(
+        all_parallel_units.len(),
+        (config.compute_nodes - 1) * config.compute_node_cores
+    );
+
+    cluster
+        .simple_restart_nodes(vec!["compute-2".to_string()])
+        .await;
+
+    sleep(Duration::from_secs(
+        MAX_HEARTBEAT_INTERVAL_SECS_CONFIG_FOR_AUTO_SCALE * 2,
+    ))
+    .await;
+
+    let table_mat_fragment = cluster
+        .locate_one_fragment(vec![
+            identity_contains("materialize"),
+            no_identity_contains("simpleAgg"),
+        ])
+        .await?;
+
+    let (all_parallel_units, used_parallel_units) = table_mat_fragment.parallel_unit_usage();
+
+    assert_eq!(all_parallel_units.len(), used_parallel_units.len());
+
+    assert_eq!(
+        all_parallel_units.len(),
+        config.compute_nodes * config.compute_node_cores
+    );
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_no_parallelism_control_with_fixed_and_auto() -> Result<()> {
+    test_auto_parallelism_control_with_fixed_and_auto_helper(false).await
+}
+
+#[tokio::test]
+async fn test_auto_parallelism_control_with_fixed_and_auto() -> Result<()> {
+    test_auto_parallelism_control_with_fixed_and_auto_helper(true).await
+}
+
+async fn test_auto_parallelism_control_with_fixed_and_auto_helper(
+    enable_auto_parallelism_control: bool,
+) -> Result<()> {
+    let config = Configuration::for_auto_parallelism(
+        MAX_HEARTBEAT_INTERVAL_SECS_CONFIG_FOR_AUTO_SCALE,
+        true,
+        enable_auto_parallelism_control,
+    );
+    let mut cluster = Cluster::start(config.clone()).await?;
+    let mut session = cluster.start_session();
+
+    // Keep one worker reserved for adding later.
+    let select_worker = "compute-2";
+    cluster
+        .simple_kill_nodes(vec![select_worker.to_string()])
+        .await;
+
+    sleep(Duration::from_secs(
+        MAX_HEARTBEAT_INTERVAL_SECS_CONFIG_FOR_AUTO_SCALE * 2,
+    ))
+    .await;
+
+    session.run("create table t (v1 int);").await?;
+
+    session
+        .run("select parallelism from rw_table_fragments")
```

Comment:
Better to define some const SQLs for all these ones, then you can use them instead.